### PR TITLE
Adding Support for GLM4

### PR DESCRIPTION
**Pull request by Goekdeniz-Guelmez**

### CODE REVIEW

**Comment by Goekdeniz-Guelmez** [comment]
```text
python -m mlx_lm.generate --model THUDM/GLM-4-9B-0414 --prompt "Write me a function that computes fibonacci in Rust" -m 128
<frozen runpy>:128: RuntimeWarning: 'mlx_lm.generate' found in sys.modules after import of package 'mlx_lm', but prior to execution of 'mlx_lm.generate'; this may result in unpredictable behaviour
Calling `python -m mlx_lm.generate...` directly is deprecated. Use `mlx_lm.generate...` or `python -m mlx_lm generate ...` instead.
Fetching 10 files: 100%|█████████████████████████████████████████████████████| 10/10 [00:00<00:00, 107546.26it/s]
==========

Certainly! Below is a Rust function that computes the Fibonacci sequence. The function takes an integer `n` and returns the `n`-th Fibonacci number.

```rust
fn fibonacci(n: u32) -> u32 {
    let mut a = 0;
    let mut b = 1;
    if n == 0 {
        return a;
    } else if n == 1 {
        return b;
    } else {
        let mut i = 2;
        while i <= n {
            let c = a + b;
            a = b;
            b = c;
            i += 1
==========
Prompt: 14 tokens, 34.109 tokens-per-sec
Generation: 128 tokens, 5.443 tokens-per-sec
Peak memory: 19.142 GB
```

**Comment by Goekdeniz-Guelmez** [comment]
Is ready to Merge!

**Comment by Goekdeniz-Guelmez** [comment]
However, the model exhibits high loss on quants, upon further investigation, I found this [Discussion](https://huggingface.co/THUDM/GLM-4-9B-0414/discussions/5) has the same problems I phase using the quants and why directly from the developers of that model.

**Comment by awni** [comment]
Looks great, thanks for the addition! Nice and simple. Could you add the model type to the list for LoRA training [here](https://github.com/ml-explore/mlx-lm/blob/main/mlx_lm/tuner/utils.py#L80)? Then we can merge it.

**Comment by Goekdeniz-Guelmez** [comment]
Ahh my bad:

```text
(mlx-dev) (base)  ~/Desktop/mlx-lm/ [adding-glm4*] python -m mlx_lm.lora \
    --model /Volumes/T7_Shield/MODELS/MLX/GLM-4-9B-0414-4bit \
    --train \
    --data /Users/gokdenizgulmez/Library/Mobile\ Documents/com\~apple\~CloudDocs/Datastes/MLX/data_smoll \
    --iters 10 \
    --batch-size 1 \
    --num-layers 1 \
    --val-batches 1 \
    --steps-per-report 1 \
    --adapter-path /Users/gokdenizgulmez/Library/Mobile\ Documents/com\~apple\~CloudDocs/Datastes/MLX/glm4-test \
    --max-seq-length 8
Calling `python -m mlx_lm.lora...` directly is deprecated. Use `mlx_lm.lora...` or `python -m mlx_lm lora ...` instead.
Loading pretrained model
Loading datasets
Training
Trainable parameters: 0.001% (0.100M/9400.279M)
Starting training..., iters: 10
[WARNING] Some sequences are longer than 8 tokens. The longest sentence 1243 will be truncated to 8. Consider pre-splitting your data to save memory.
[WARNING] Some sequences are longer than 8 tokens. The longest sentence 1958 will be truncated to 8. Consider pre-splitting your data to save memory.
Iter 1: Val loss 2.407, Val took 0.396s
Iter 1: Train loss 2.407, Learning Rate 1.000e-05, It/sec 0.808, Tokens/sec 5.656, Trained Tokens 7, Peak mem 5.306 GB
[WARNING] Some sequences are longer than 8 tokens. The longest sentence 659 will be truncated to 8. Consider pre-splitting your data to save memory.
Iter 2: Train loss 2.401, Learning Rate 1.000e-05, It/sec 3.615, Tokens/sec 25.308, Trained Tokens 14, Peak mem 5.312 GB
[WARNING] Some sequences are longer than 8 tokens. The longest sentence 1887 will be truncated to 8. Consider pre-splitting your data to save memory.
Iter 3: Train loss 2.391, Learning Rate 1.000e-05, It/sec 4.833, Tokens/sec 33.834, Trained Tokens 21, Peak mem 5.312 GB
[WARNING] Some sequences are longer than 8 tokens. The longest sentence 1700 will be truncated to 8. Consider pre-splitting your data to save memory.
Iter 4: Train loss 2.373, Learning Rate 1.000e-05, It/sec 5.071, Tokens/sec 35.500, Trained Tokens 28, Peak mem 5.312 GB
[WARNING] Some sequences are longer than 8 tokens. The longest sentence 1444 will be truncated to 8. Consider pre-splitting your data to save memory.
Iter 5: Train loss 2.352, Learning Rate 1.000e-05, It/sec 5.052, Tokens/sec 35.366, Trained Tokens 35, Peak mem 5.312 GB
[WARNING] Some sequences are longer than 8 tokens. The longest sentence 1119 will be truncated to 8. Consider pre-splitting your data to save memory.
Iter 6: Train loss 2.327, Learning Rate 1.000e-05, It/sec 5.093, Tokens/sec 35.649, Trained Tokens 42, Peak mem 5.312 GB
[WARNING] Some sequences are longer than 8 tokens. The longest sentence 944 will be truncated to 8. Consider pre-splitting your data to save memory.
Iter 7: Train loss 2.299, Learning Rate 1.000e-05, It/sec 5.059, Tokens/sec 35.410, Trained Tokens 49, Peak mem 5.312 GB
[WARNING] Some sequences are longer than 8 tokens. The longest sentence 1749 will be truncated to 8. Consider pre-splitting your data to save memory.
Iter 8: Train loss 2.265, Learning Rate 1.000e-05, It/sec 5.046, Tokens/sec 35.323, Trained Tokens 56, Peak mem 5.312 GB
[WARNING] Some sequences are longer than 8 tokens. The longest sentence 1381 will be truncated to 8. Consider pre-splitting your data to save memory.
Iter 9: Train loss 2.228, Learning Rate 1.000e-05, It/sec 5.076, Tokens/sec 35.534, Trained Tokens 63, Peak mem 5.312 GB
[WARNING] Some sequences are longer than 8 tokens. The longest sentence 791 will be truncated to 8. Consider pre-splitting your data to save memory.
[WARNING] Some sequences are longer than 8 tokens. The longest sentence 1113 will be truncated to 8. Consider pre-splitting your data to save memory.
Iter 10: Val loss 2.182, Val took 0.183s
Iter 10: Train loss 2.182, Learning Rate 1.000e-05, It/sec 5.067, Tokens/sec 35.469, Trained Tokens 70, Peak mem 5.312 GB
Saved final weights to /Users/gokdenizgulmez/Library/Mobile Documents/com~apple~CloudDocs/Datastes/MLX/glm4-test/adapters.safetensors.
```

**Review by awni — APPROVED** [review]
Thanks!!

**Comment by Goekdeniz-Guelmez** [comment]
Always a pleasure!

**Comment by Goekdeniz-Guelmez** [comment]
Hey @awni, I think this hasn't been merged, probably due to me closing it too soon :D
